yarpl/include/yarpl/Refcounted.h

```diff
@@ -311,56 +312,49 @@ class AtomicReference final {
     dec();
   }
 
-  template <typename U>
-  AtomicReference& operator=(Reference<U>&& other) {
-    return assign(std::move(other));
-  }
-
   Reference<T> load() {
-    return Reference<T>(*this);
+    std::lock_guard<std::mutex> lg{mtx_};
+    if (pointer_) {
+      pointer_->incRef();
+      Reference<T> r{pointer_, detail::skip_initial_refcount_check{}};
+      return r;
+    }
+    else {
+      Reference<T> r{nullptr};
+      return r;
+    }
   }
 
   Reference<T> exchange(Reference<T> other) {
-    T* old = pointer_.exchange(other.get());
+    std::lock_guard<std::mutex> lg{mtx_};
+    T* old = pointer_;
+    pointer_ = other.get();
     inc();
     Reference<T> r{old, detail::skip_initial_refcount_check{}};
     return r;
   }
 
-  void store(Reference<T> const& other) {
+  void store(Reference<T> const& other) {   
+    std::lock_guard<std::mutex> lg{mtx_};
     dec();
-    pointer_.store(other.get());
+    pointer_ = other.pointer_;
     inc();
   }
 
-  explicit operator bool() const {
-    return pointer_;
-  }
-
-  T* operator->() const {
-    return pointer_;
+  void store(std::nullptr_t) {
```

Comment:
Why is this needed?  Is there not an implicit `Reference<T>` ctor from `std::nullptr_t`?

Reply:
Just an optimization, same as the Reference<> case. It would also be just excellent if we could get rid of Reference and avoid all this mess.